### PR TITLE
Core/Movement: Add a timed event to respawn the creature on HomeMovementGenerator

### PR DESCRIPTION
**Pull request by ccrs**
**Changes proposed**:

If the evading creature cannot find a valid path towards home, relocate it in 3 seconds.

**Target branch(es)**: 335/6x

I've separated this part from the main PR as it introduces a new mechanic.
Looking forward to some nice and juicy feedback <3

### CODE REVIEW

File: src/server/game/AI/CreatureAI.h [hunk 1]
```diff
@@ -218,4 +219,51 @@ enum Permitions
...
 };
 
+class EvadeHomeCall : public BasicEvent
+{
+public:
+    explicit EvadeHomeCall(Creature* owner) : _owner(owner) { }
+
+    bool Execute(uint64 /*time*/, uint32 /*diff*/)
+    {
+        if (_owner->IsAIEnabled && _owner->GetAI() && reinterpret_cast<const SmartAI*>(_owner->GetAI()) == nullptr)
```
Comment (DDuarte) [inline src/server/game/AI/CreatureAI.h]: I guess this should be a dynamic_cast, not a reinterpret one. That's evil, very evil.

Comment (ccrs) [inline src/server/game/AI/CreatureAI.h]: SmartAI uses JustReachedHome on respawn while the rest don't :D

File: src/server/game/AI/CreatureAI.h [hunk 2]
```diff
@@ -218,4 +219,51 @@ enum Permitions
...
+};
+
+class EvadeDespawner : public BasicEvent
+{
+public:
+    explicit EvadeDespawner(Creature* owner) : _owner(owner) { }
+
+    bool Execute(uint64 /*time*/, uint32 /*diff*/)
+    {
+        _owner->Respawn(true);
```
Comment (ccrs) [inline src/server/game/AI/CreatureAI.h]: Maybe avoid this part? There is no other way to respawn summons...

File: src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h [hunk 3]
```diff
@@ -20,18 +20,17 @@
...
+template <class T>
 class HomeMovementGenerator;
 
 template <>
 class HomeMovementGenerator<Creature> : public MovementGeneratorMedium< Creature, HomeMovementGenerator<Creature> >
 {
     public:
 
-        HomeMovementGenerator() : arrived(false) { }
+        HomeMovementGenerator() : m_arrived(false), m_travelInitialized(false), m_despawnTimer(30 * IN_MILLISECONDS) { }
```
Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h]: Make the timer longer? Make a CreatureAI function to change it on demand?

File: src/server/game/AI/CreatureAI.h (hunk 1 above)
Comment (ccrs) [inline src/server/game/AI/CreatureAI.h]: dynamic_cast forces the header include, and eitherway I'm not actually using the casted pointer.

Comment (Shauren) [inline src/server/game/AI/CreatureAI.h]: use AIName from template then

Comment (ccrs) [inline src/server/game/AI/CreatureAI.h]: There are CreatureScripts using SmartAI, aren't they?

Comment (Shauren) [inline src/server/game/AI/CreatureAI.h]: then move this method to .cpp and include smartai
reinterpret_cast doesn't do what you want here - it does NOT return nullptr when the cast is invalid, only dynamic_cast does that (and to know whether the cast is valid it needs the class definition, that is why you need the include)

**Comment by Nawuko** [comment]
Maybe we should summon @r00ty-tc I dont know if this changes collide with the dynamic branch due to use of the Respawn(true) function

**Comment by Trisjdc** [comment]
What's the point of this?

**Comment by ccrs** [comment]
Easy: ````quote If the evading creature has not reached home in 30 seconds, respawn it.
```` Example: pathgenerator cant find a valid path to home and creature gets stuck forever in combat.

**Comment by Nawuko** [comment]
@ccrs But getting stuck forever is retail behavior xD

**Comment by ccrs** [comment]
It's not if the stuck is caused by creature not finding a valid path. In that case the creature just despawns after some time. The only case in which creatures got stuck on evade on retail (older versions) was on instances if the target wasn't reachable.

**Comment by Trisjdc** [comment]
It would be better to just force the underlying pathfinding to _always_ return a path that ends at home position, at least for evade, than to risk doing a weird despawn in case the return path takes too long.

**Comment by ccrs** [comment]
Yeah, as u posted ur comment I realized exactly the same.
I've already done the modifications, tomorrow I'll update this :)

**Comment by Treeston** [comment]
@ccrs what happened to you updating this tomorrow?

**Comment by Aokromes** [comment]
Any reason to don't merge this?

**Comment by ccrs** [comment]
Still waiting for some random member approbation.

**Comment by Aokromes** [comment]
If there is no reason to merge this, i will merge it on 24h.

**Comment by joschiwald** [comment]
where are these 30 seconds in your code?

File: src/server/game/Movement/MovementGenerators/HomeMovementGenerator.cpp [hunk 4]
```diff
@@ -22,48 +22,79 @@
...
-    return !arrived;
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE))
+        return true;
+
+    if (!m_travelInitialized)
+        SetTargetLocation(owner);
+
+    if (m_despawnEvent)
+    {
+        owner->m_Events.AddEvent(new EvadeDespawner(owner), owner->m_Events.CalculateTime(3000));
```
Comment (Demonid) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.cpp]: Inst this 30000?

Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.cpp]: nah, I didnt update the description.

**Comment by ccrs** [comment]
@joschiwald description updated.

File: src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h [hunk 5]
```diff
@@ -21,18 +21,15 @@
 
 #include "MovementGenerator.h"
 
-class Creature;
```
Comment (jackpoz) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h]: why was this removed ?

Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h]: ?
because its not needed?

Comment (jackpoz) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h]: and why it's not needed ? could you please elaborate more in detail why it was removed and why it shouldn't be needed ?

**Comment by Aokromes** [comment]
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.

File: src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h (hunk 5 above)
Comment (ccrs) [inline src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h]: I guess someone thought it was needed to declare the class for the template defines or w/e, but the thing is its not needed. Wether its because the MovementGenerator.h include or the class inherit I dont care, just checked its not needed.

**Comment by ccrs** [comment]
Meh I will

**Comment by Aokromes** [comment]
@ccrs 
Details continuous-integration/travis-ci/pr — The Travis CI build passed
This branch has conflicts that must be resolved
 Use the command line to resolve conflicts before continuing.
Conflicting files
src/server/game/Movement/MovementGenerators/HomeMovementGenerator.cpp
src/server/game/Movement/MovementGenerators/HomeMovementGenerator.h
